h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -225,4 +227,92 @@ public void close() throws IOException {
       }
     };
   }
+
+  @Override
+  public OutputStream create(String path, boolean overwrite) {
+    final GcsBlob gcsBlob = GcsBlob.of(path);
+    Log.debug("Creating: " + gcsBlob.getCanonical());
+    final WriteChannel writer = storage.create(gcsBlob.getBlobInfo()).writer();
+    return new OutputStream() {
+      @Override
+      public void write(int b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(new byte[]{(byte) b});
+        writer.write(buffer);
+      }
+
+      @Override
+      public void write(byte[] b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(b);
+        writer.write(buffer);
+      }
+
+      @Override
+      public void write(byte[] b, int off, int len) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(b, off, len);
+        writer.write(buffer);
+      }
+
+      @Override
+      public void close() throws IOException {
+        writer.close();
+      }
+    };
+  }
+
+  @Override
+  public boolean rename(String fromPath, String toPath) {
+    final BlobId fromBlob = GcsBlob.of(fromPath).getBlobId();
+    final BlobId toBlob = GcsBlob.of(toPath).getBlobId();
+
+    storage.get(fromBlob).copyTo(toBlob);
+    keyCache.invalidate(fromBlob.getBucket());
+    keyCache.invalidate(toBlob.getBucket());
+    return storage.delete(fromBlob);
+  }
+
+  @Override
+  public boolean exists(String path) {
+    final String bk[] = GcsBlob.removePrefix(path).split("/", 2);
+    if (bk.length == 1) {
+      return storage.get(bk[0]).exists();
+    } else if (bk.length == 2) {
+      return storage.get(bk[0], bk[1]).exists();
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public boolean delete(String path) {
+    final BlobId blob = GcsBlob.of(path).getBlobId();
+    keyCache.invalidate(blob.getBucket());
+    return storage.get(blob).delete();
+  }
+
+  @Override
+  public long length(String path) {
+    final BlobId blob = GcsBlob.of(path).getBlobId();
+    return storage.get(blob).getSize();
+  }
+
+  /**
+   * Lists Blobs prefixed with `path`.
+   * Prefix `path` is removed from the name of returned entries.
+   * e.g.
+   * If `path` equals gs://bucket/infix and 2 Blobs exist: "gs://bucket/infix/blob1, gs://bucket/infix/blob2,
+   * the returned array contains of Persist Entries with names set to blob1 and blob2, respectively.
+   */
+  @Override
+  public PersistEntry[] list(String path) {
+    List<String> matches = calcTypeaheadMatches(path, Integer.MAX_VALUE);
```

Comment:
I would avoid using the typeahead cache here. Also not that current implementation of `calcTypeaheadMatches` doesnt't actually use the `limit` parameter

Reply:
Good points!
Resolved both.
See 3ac6e31 and bcb35bf.